src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,327 @@
+    convolution.
+  - `first_dilation`: the dilation of the 3x3 convolution.
+  - `activation`: the activation function to use.
+  - `connection`: the function applied to the output of residual and skip paths in
+    a block. See [`addact`](#) and [`actadd`](#) for an example. Note that this uses
+    PartialFunctions.jl to pass in the activation function with the notation `addact\$activation`.
+  - `norm_layer`: the normalization layer to use.
+  - `drop_block`: the drop block layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `drop_path`: the drop path layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `attn_fn`: the attention function to use. See [`squeeze_excite`](#) for an example.
+"""
+function bottleneck(inplanes, planes; stride = 1, cardinality = 1, base_width = 64,
+                    reduction_factor = 1, activation = relu,
+                    norm_layer = BatchNorm, prenorm = false,
+                    drop_block = identity, drop_path = identity,
+                    attn_fn = planes -> identity)
+    width = floor(Int, planes * (base_width / 64)) * cardinality
+    first_planes = width ÷ reduction_factor
+    outplanes = planes * expansion_factor(bottleneck)
+    conv_bn1 = conv_norm((1, 1), inplanes => first_planes, activation; norm_layer, prenorm,
+                         bias = false)
+    conv_bn2 = conv_norm((3, 3), first_planes => width, identity; norm_layer, prenorm,
+                         stride, pad = 1, groups = cardinality, bias = false)
+    conv_bn3 = conv_norm((1, 1), width => outplanes, identity; norm_layer, prenorm,
+                         bias = false)
+    layers = [conv_bn1..., conv_bn2..., drop_block, activation, conv_bn3...,
+        attn_fn(outplanes), drop_path]
+    return Chain(filter!(!=(identity), layers)...)
+end
+expansion_factor(::typeof(bottleneck)) = 4
+
+# Downsample layer using convolutions.
+function downsample_conv(inplanes::Integer, outplanes::Integer; stride::Integer = 1,
+                         norm_layer = BatchNorm, prenorm = false)
+    return Chain(conv_norm((1, 1), inplanes => outplanes, identity; norm_layer, prenorm,
+                           pad = SamePad(), stride, bias = false)...)
+end
+
+# Downsample layer using max pooling
+function downsample_pool(inplanes::Integer, outplanes::Integer; stride::Integer = 1,
+                         norm_layer = BatchNorm, prenorm = false)
+    pool = (stride == 1) ? identity : MeanPool((2, 2); stride, pad = SamePad())
+    return Chain(pool,
+                 conv_norm((1, 1), inplanes => outplanes, identity; norm_layer, prenorm,
+                           bias = false)...)
+end
+
+# Downsample layer which is an identity projection. Uses max pooling
+# when the output size is more than the input size.
+function downsample_identity(inplanes::Integer, outplanes::Integer; kwargs...)
```

Comment:
Again, necessary. The other downsamples have kwargs, and we can't filter based on the type of downsample

Reply:
Umm....not _yet_, actually. I did away with templating so this is a weird outstanding issue. I think a `downsample_builder` is the only option I see

Reply:
Not sure how to handle this, though. This is a surprisingly annoying edge case, since only one of the three downsample functions doesn't correspond to similar keyword arguments.

Reply:
Yeah, we can leave this as clean up for another PR.